### PR TITLE
PERF-#5017: `reset_index` shouldn't trigger index materialization if possible

### LINKED ISSUE #5017: PERF: `reset_index` shouldn't trigger index materialization in case when `drop==True` and `level==None`
RangeIndex can be built on the basis of `_row_lengths`. Materializing lengths is much cheaper than materializing indices.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5017 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by mvashishtha — APPROVED** [review]

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 1]
```diff
@@ -627,7 +627,8 @@ def reset_index(self, **kwargs):
         else:
             new_self = self.copy()
             new_self.index = (
-                pandas.RangeIndex(len(new_self.index))
+                # Cheaper to compute row lengths than index
+                pandas.RangeIndex(sum(new_self._modin_frame._row_lengths))
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:631]: Do we really want to make QC know that `_modin_frame` has `_row_lengths`? Same question got hung in #4460.

Comment (YarShev) [reply modin/core/storage_formats/pandas/query_compiler.py:631]: cc @mvashishtha, @devin-petersohn, @RehanSD, @vnlitvinov

Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:631]: @YarShev Why not? The modin is based on this mechanism, in addition, this information can be used to advantage in some functions. I think that we can make these functions public (`_row_lengths/_column_widths`).

Comment (YarShev) [reply modin/core/storage_formats/pandas/query_compiler.py:631]: I am okay with this ([link](https://github.com/modin-project/modin/pull/4460#discussion_r876251431)) but @devin-petersohn, @mvashishtha might have different thoughts on the matter. I would like to make sure we are on the same page.

Comment (mvashishtha) [reply modin/core/storage_formats/pandas/query_compiler.py:631]: I support making `row_lengths` and `column_widths` public.

Comment (YarShev) [reply modin/core/storage_formats/pandas/query_compiler.py:631]: Good, let's make those public. Please create an issue for that.

**Review by YarShev — APPROVED** [review]

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 1 above)
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:631]: https://github.com/modin-project/modin/issues/5024

**Pull request merged**